sdk/maps/azure-maps-geolocation/src/test/java/com/azure/maps/geolocation/GeolocationClientTestBase.java

```diff
@@ -3,95 +3,114 @@
     }
 
-    GeolocationClientBuilder getGeoLocationAsyncClientBuilder(HttpClient httpClient, GeolocationServiceVersion serviceVersion) {
+    GeolocationClientBuilder getGeoLocationAsyncClientBuilder(HttpClient httpClient,
+                                                              GeolocationServiceVersion serviceVersion) {
         GeolocationClientBuilder builder = new GeolocationClientBuilder()
+            .pipeline(getHttpPipeline(httpClient))
             .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS))
             .serviceVersion(serviceVersion);
-        String endpoint = getEndpoint();
-        if (getEndpoint() != null) {
-            builder.endpoint(endpoint);
-        }
-        if (getTestMode() == TestMode.RECORD) {
-            builder.addPolicy(interceptorManager.getRecordPolicy());
-        }
-        if (getTestMode() == TestMode.PLAYBACK) {
-            builder.credential(new AzureKeyCredential(FAKE_API_KEY)).httpClient(interceptorManager.getPlaybackClient());
-        } else {
-            builder.credential((new AzureKeyCredential(
-                Configuration.getGlobalConfiguration().get("SUBSCRIPTION_KEY"))));
+
+        if (interceptorManager.isPlaybackMode()) {
+            builder.endpoint("https://localhost:8080");
         }
+
         return builder;
     }
 
     HttpPipeline getHttpPipeline(HttpClient httpClient) {
-        TokenCredential credential = null;
+        httpClient = interceptorManager.isPlaybackMode() ? interceptorManager.getPlaybackClient() : httpClient;
 
-        if (!interceptorManager.isPlaybackMode()) {
-            credential = new EnvironmentCredentialBuilder().httpClient(httpClient).build();
+        if (interceptorManager.isRecordMode()) {
+            interceptorManager.addSanitizers(
+                Collections.singletonList(
+                    new TestProxySanitizer("subscription-key", ".+", "REDACTED", TestProxySanitizerType.HEADER)));
         }
 
-        final List<HttpPipelinePolicy> policies = new ArrayList<>();
-        if (credential != null) {
-            policies.add(new BearerTokenAuthenticationPolicy(credential, endpoint.replaceFirst("/$", "") + "/.default"));
+        if (interceptorManager.isPlaybackMode()) {
+            List<TestProxyRequestMatcher> customMatchers = new ArrayList<>();
+
+            customMatchers.add(new CustomMatcher().setExcludedHeaders(Collections.singletonList("subscription-key")));
+            interceptorManager.addMatchers(customMatchers);
         }
 
-        if (getTestMode() == TestMode.RECORD) {
+        final List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        policies.add(new UserAgentPolicy(null, SDK_NAME, SDK_VERSION, Configuration.getGlobalConfiguration().clone()));
+
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(new RetryPolicy(new ExponentialBackoff(5, Duration.ofSeconds(2), Duration.ofSeconds(16))));
```

Comment:
Shouldn't this be done in the builder?

Reply:
We can do that, I just followed a pattern we also have on Key Vault. I'll try to use the builder instead.

Reply:
I decided to keep things as-is, it's easier to use `HttpPolicyProviders.addBeforeRetryPolicies()` and `HttpPolicyProviders.addAfterRetryPolicies()`, as well as `HttpPipelineBuilder.policies()` method to add all the policies that will come into effect. Else, I would have to do that one by one with `GeolocationClientBuilder.addPolicy()`.